### PR TITLE
Added unstable-schema generation for Cargo.toml

### PR DESCRIPTION
**Pull request by dacianpascu06**
### What does this PR try to resolve?

Added unstable-schema feature that generates JsonSchema for Cargo.toml

See #12883
finished first step of [plan](https://github.com/rust-lang/cargo/issues/12883#issuecomment-2407648385)

### Information
In cargo-util-schemas,  run cargo test --feature unstable-schema  . If there have been any changes to manifest.schema.json, rerun with env variable SNAPSHOTS=overwrite  and it will update them.

### How should we test and review this PR?
In cargo-util-schemas run cargo test --features unstable-schema and it will generate manifest.schema.json

### CODE REVIEW

File: crates/cargo-util-schemas/Cargo.toml [hunk 1]
```diff
@@ -17,9 +17,14 @@ thiserror.workspace = true
 toml.workspace = true
 unicode-xid.workspace = true
 url.workspace = true
+schemars = { version = "0.8.21", features = ["preserve_order","semver"], optional = true }
+serde_json = { version = "1.0", optional = true }
```
Comment (weihanglo) [inline crates/cargo-util-schemas/Cargo.toml:21]: In Cargo we use [workspace inheritances](https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#inheriting-a-dependency-from-a-workspace) as much as possible. You may need to do that as well, i.e,

```suggestion
schemars = { workspace = true, features = ["preserve_order","semver"], optional = true }
serde_json = { workspace = true, optional = true }
```

Could you rearrange dependencies in alphabetically order?
(Sorry that we don't have an automatic mechanism doing to this now 😅)

File: crates/cargo-util-schemas/schema.json
Comment (weihanglo) [inline crates/cargo-util-schemas/schema.json:1]: Could we have more info in the PR description about:

* How to generate this schema.json?
* How to review this schema.json file? 
* Assumedly [this plan](https://github.com/rust-lang/cargo/issues/12883#issuecomment-2407648385) is followed. Which tasks are done and which are still pending in this PR?

[the remark this answers is not in the document]
Comment (dacianpascu06) [reply crates/cargo-util-schemas/schema.json:1]: After last commit :  snapbox::assert_data_eq!(dump, snapbox::file!["../schema.json"]); checks if there have been changes.

**Comment by epage** [comment]
Please cleanup the commits for how you'd like them reviewed / merged.

File: crates/cargo-util-schemas/src/core/partial_version.rs [hunk 2]
```diff
@@ -4,11 +4,14 @@ use semver::{Comparator, Version, VersionReq};
 use serde_untagged::UntaggedEnumVisitor;
 
 #[derive(PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Debug)]
+#[cfg_attr(feature = "unstable-schema", derive(schemars::JsonSchema))]
```
Comment (epage) [inline crates/cargo-util-schemas/src/core/partial_version.rs:7]: Is this still needed?

File: crates/cargo-util-schemas/src/schema.rs [hunk 3]
```diff
@@ -0,0 +1,68 @@
+#[allow(unused_imports)]
+use crate::manifest::TomlManifest;
+#[test]
+#[cfg(feature = "unstable-schema")]
```
Comment (epage) [inline crates/cargo-util-schemas/src/schema.rs:4]: put a cfg on this entire mod

File: crates/cargo-util-schemas/src/core/partial_version.rs (hunk 2 above)
Comment (dacianpascu06) [reply crates/cargo-util-schemas/src/core/partial_version.rs]: removed it , thanks

File: crates/cargo-util-schemas/src/schema.rs (hunk 3 above)
Comment (dacianpascu06) [reply crates/cargo-util-schemas/src/schema.rs]: ✅

File: crates/cargo-util-schemas/schema.json
Comment (dacianpascu06) [reply crates/cargo-util-schemas/schema.json:1]: Done.

File: crates/cargo-util-schemas/src/manifest/mod.rs [hunk 4]
```diff
@@ -18,16 +20,19 @@ use serde_untagged::UntaggedEnumVisitor;
 
 use crate::core::PackageIdSpec;
 use crate::restricted_names;
-
```
Comment (epage) [inline crates/cargo-util-schemas/src/manifest/mod.rs:21]: Why was this line removed

File: crates/cargo-util-schemas/src/manifest/mod.rs [hunk 5]
```diff
@@ -1592,6 +1635,7 @@ pub enum TomlLintLevel {
 }
 
 #[derive(Copy, Clone, Debug)]
+#[cfg_attr(feature = "unstable-schema", derive(schemars::JsonSchema))]
 pub struct InvalidCargoFeatures {}
```
Comment (epage) [inline crates/cargo-util-schemas/src/manifest/mod.rs:1639]: This should be skipped rather than be in the xhema

File: crates/cargo-util-schemas/src/manifest/mod.rs [hunk 6]
```diff
@@ -157,9 +167,12 @@ pub struct InheritablePackage {
 /// tables.
 #[derive(Deserialize, Serialize, Clone, Debug)]
 #[serde(rename_all = "kebab-case")]
+#[cfg_attr(feature = "unstable-schema", derive(schemars::JsonSchema))]
 pub struct TomlPackage {
     pub edition: Option<InheritableString>,
+    #[cfg_attr(feature = "unstable-schema", schemars(with = "String"))]
     pub rust_version: Option<InheritableRustVersion>,
```
Comment (epage) [inline crates/cargo-util-schemas/src/manifest/mod.rs:174]: These attributes aren't acknowledging the `None` case

File: crates/cargo-util-schemas/Cargo.toml (hunk 1 above)
Comment (epage) [reply crates/cargo-util-schemas/Cargo.toml]: Don't forget to do this

File: crates/cargo-util-schemas/schema.json
[the remark this answers is not in the document]
Comment (epage) [reply crates/cargo-util-schemas/schema.json:1]: I see snapshotting it into the crate source is a starting point.  As we explore how to be exposing this, we can see where it will work to be.  I'm unsure if a json file in the docs would be best but I hope we can eventually code generate the example blocks in the docs.

File: crates/cargo-util-schemas/src/manifest/mod.rs (hunk 6 above)
Comment (dacianpascu06) [reply crates/cargo-util-schemas/src/manifest/mod.rs]: fixed this for more types

File: crates/cargo-util-schemas/src/manifest/mod.rs (hunk 5 above)
Comment (dacianpascu06) [reply crates/cargo-util-schemas/src/manifest/mod.rs]: deleted this line as invalidcargofeatures is skipped anyway

File: crates/cargo-util-schemas/schema.json
[the remark this answers is not in the document]
Comment (weihanglo) [reply crates/cargo-util-schemas/schema.json:1]: > json file in the docs

To clarify, I meant an asset published aling with docs, so it can be access via some URL like <https://doc.rust-lang.org/nightly/cargo/manifest-schema.json>.

[the remark this answers is not in the document]
Comment (epage) [reply crates/cargo-util-schemas/schema.json:1]: Using our crate policy terms, I would describe this as Experimental at this time.  Based on our the stability we shoot for with our book, it seems too early to post this there.  Will we want to continue to produce this?  How would people want to access this?

File: crates/cargo-util-schemas/src/schema.rs [hunk 7]
```diff
@@ -0,0 +1,56 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::string::String;
+use toml::Value as TomlValue;
+#[derive(Serialize, Deserialize)]
+pub struct TomlValueWrapper(pub TomlValue);
+#[test]
+fn dump_schema() {
```
Comment (epage) [inline crates/cargo-util-schemas/src/schema.rs:9]: Generally you want a newline between `use`, types, and functions.

I would not put an unrelated function between the type and its impl block

Comment (epage) [reply crates/cargo-util-schemas/src/schema.rs:9]: Generally, I would expect a test to be at the bottom of a file.

In this case, I think `dump_schema` should likely live in
- `manifest/mod.rs` as its dumping that type
- an integration test for schemas in a function named for dealing with the manifest schema

File: crates/cargo-util-schemas/src/schema.rs [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+    let schema = schemars::schema_for!(crate::manifest::TomlManifest);
+    let dump = serde_json::to_string_pretty(&schema).unwrap();
+    snapbox::assert_data_eq!(dump, snapbox::file!("../schema.json"));
```
Comment (epage) [inline crates/cargo-util-schemas/src/schema.rs:12]: We're going to have more than one schema file.  We should qualify which schema this is

File: crates/cargo-util-schemas/Cargo.toml (hunk 1 above)
Comment (dacianpascu06) [reply crates/cargo-util-schemas/Cargo.toml]: I've rearranged them but I don't know how to use workspace for optional dependencies that are active only when the feature flag is set.

Comment (epage) [reply crates/cargo-util-schemas/Cargo.toml]: Only inherit the version. Everything else will work the same.

Comment (dacianpascu06) [reply crates/cargo-util-schemas/Cargo.toml]: ✅

File: crates/cargo-util-schemas/Cargo.toml [hunk 9]
```diff
@@ -9,8 +9,10 @@ repository.workspace = true
 description = "Deserialization schemas for Cargo"
 
 [dependencies]
+schemars = { workspace = true, features = ["preserve_order","semver"], optional = true }
 semver.workspace = true
 serde = { workspace = true, features = ["derive"] }
+serde_json = { workspace = true, optional = true }
```
Comment (epage) [inline crates/cargo-util-schemas/Cargo.toml:15]: This will cause `Cargo.lock` to change and that change needs to be included to pass CI

File: crates/cargo-util-schemas/Cargo.toml
Comment (epage) [inline crates/cargo-util-schemas/Cargo.toml:1]: Looks like `cargo-util-schemas` `package.version` needs to be updated

**Comment by epage** [comment]
Thanks!

@bors r+

File: Cargo.lock
Comment (weihanglo) [inline Cargo.lock:1]: @dacianpascu06 could you remove unnecessary lockfile updates from the commit?

This could generate the minimal changes we need:

```
git checkout HEAD~ -- Cargo.lock
cargo b
```

The diff should be something like:

```diff
diff --git a/Cargo.lock b/Cargo.lock
index a943b8b8c40..dd5ff342bbc 100644
--- a/Cargo.lock
+++ b/Cargo.lock
@@ -327,7 +327,7 @@ dependencies = [
  "humantime",
  "ignore",
  "im-rc",
- "indexmap",
+ "indexmap 2.3.0",
  "itertools 0.13.0",
  "jobserver",
  "lazycell",
@@ -492,10 +492,12 @@ dependencies = [
 name = "cargo-util-schemas"
 version = "0.7.1"
 dependencies = [
+ "schemars",
  "semver",
  "serde",
  "serde-untagged",
  "serde-value",
+ "serde_json",
  "snapbox",
  "thiserror",
  "toml",
@@ -902,6 +904,12 @@ version = "1.0.4"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "56ce8c6da7551ec6c462cbaf3bfbc75131ebbfa1c944aeaa9dab51ca1c5f0c3b"
 
+[[package]]
+name = "dyn-clone"
+version = "1.0.17"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "0d6ef0072f8a535281e4876be788938b528e9a1d43900b82c2569af7da799125"
+
 [[package]]
 name = "ecdsa"
 version = "0.16.9"
@@ -1466,7 +1474,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "7ddf80e16f3c19ac06ce415a38b8591993d3f73aede049cb561becb5b3a8e242"
 dependencies = [
  "gix-hash",
- "hashbrown",
+ "hashbrown 0.14.5",
  "parking_lot",
 ]
 
@@ -1502,7 +1510,7 @@ dependencies = [
  "gix-traverse",
  "gix-utils",
  "gix-validate",
- "hashbrown",
+ "hashbrown 0.14.5",
  "itoa 1.0.11",
  "libc",
  "memmap2",
@@ -1959,6 +1967,12 @@ dependencies = [
  "walkdir",
 ]
 
+[[package]]
+name = "hashbrown"
+version = "0.12.3"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "8a9ee70c43aaf417c914396645a0fa852624801b24ebb7ae78fe8272889ac888"
+
 [[package]]
 name = "hashbrown"
 version = "0.14.5"
@@ -1975,7 +1989,7 @@ version = "0.9.0"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "692eaaf7f7607518dd3cef090f1474b61edc5301d8012f09579920df68b725ee"
 dependencies = [
- "hashbrown",
+ "hashbrown 0.14.5",
 ]
 
 [[package]]
@@ -2088,6 +2102,17 @@ dependencies = [
  "version_check",
 ]
 
+[[package]]
+name = "indexmap"
+version = "1.9.3"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "bd070e393353796e801d209ad339e89596eb4c8d430d18ede6a1cced8fafbd99"
+dependencies = [
+ "autocfg",
+ "hashbrown 0.12.3",
+ "serde",
+]
+
 [[package]]
 name = "indexmap"
 version = "2.3.0"
@@ -2095,7 +2120,7 @@ source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "de3fc2e30ba82dd1b3911c8de1ffc143c74a914a14e99514d7637e3099df5ea0"
 dependencies = [
  "equivalent",
- "hashbrown",
+ "hashbrown 0.14.5",
 ]
 
 [[package]]
@@ -3077,6 +3102,32 @@ dependencies = [
  "windows-sys 0.52.0",
 ]
 
+[[package]]
+name = "schemars"
+version = "0.8.21"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "09c024468a378b7e36765cd36702b7a90cc3cba11654f6685c8f233408e89e92"
+dependencies = [
+ "dyn-clone",
+ "indexmap 1.9.3",
+ "schemars_derive",
+ "semver",
+ "serde",
+ "serde_json",
+]
+
+[[package]]
+name = "schemars_derive"
+version = "0.8.21"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "b1eee588578aff73f856ab961cd2f79e36bc45d7ded33a7562adba4667aecc0e"
+dependencies = [
+ "proc-macro2",
+ "quote",
+ "serde_derive_internals",
+ "syn 2.0.72",
+]
+
 [[package]]
 name = "scopeguard"
 version = "1.2.0"
@@ -3177,6 +3228,17 @@ dependencies = [
  "syn 2.0.72",
 ]
 
+[[package]]
+name = "serde_derive_internals"
+version = "0.29.1"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "18d26a20a969b9e3fdf2fc2d9f21eda6c40e2de84c9408bb5d3b05d499aae711"
+dependencies = [
+ "proc-macro2",
+ "quote",
+ "syn 2.0.72",
+]
+
 [[package]]
 name = "serde_ignored"
 version = "0.1.10"
@@ -3557,7 +3619,7 @@ version = "0.22.20"
 source = "registry+https://github.com/rust-lang/crates.io-index"
 checksum = "583c44c02ad26b0c3f3066fe629275e50627026c51ac2e595cca4c230ce1ce1d"
 dependencies = [
- "indexmap",
+ "indexmap 2.3.0",
  "serde",
  "serde_spanned",
  "toml_datetime",
```

**Comment by epage** [comment]
Thanks!

@bors r+
